### ISSUE TITLE
Enable `ParallelGroup` resource

### ISSUE THREAD

**Issue by SteveL-MSFT** [PowerShell/DSC#42]
### Summary of the new feature / enhancement

Enable `-parallel` switch for `dsc`

**Comment by SteveL-MSFT** [comment]
I'm seeing more people asking for parallel/concurrency.  My original concern is it puts the onus on the configuration author to determine what can be safely run concurrently.  Any race conditions are hard to detect and when they hit, it's not easy to diagnose what the problem was nor what the result would be if there isn't a failure.

I think it would be difficult for resources to declare `canParallel` since they don't know what other resources may be running at the same time.  From their perspective, they could mean that multiple instances of the same resource can be run in parallel, but it doesn't mean that resource can run in parallel with any arbitrary resource.

So I think we can take this as a two-prong approach.  Resources can declare themselves `concurrentSafe` (or something like that) and the engine can optimize execution to run those in parallel.  Separately, we can also add a new directive `concurrentSafe` that allows the engine to run those as declared by the configuration author in parallel.

**Comment by michaeltlombardi** [comment]
My personal preference would be for DSC to define the contract for how _resources_ indicate that they are safe for concurrent operations but leave actual parallelization concerns to higher order tools (HOTs).

I'm wary of building it directly into the engine for all the edge case and race condition scenarios that will be blamed on DSC itself. In particular, I don't really like the idea of quietly modifying the execution graph based on concurrency indicators from the resources - I would much prefer that resource instances must be intentionally grouped for concurrent operations _in the configuration document_ and DSC validate that the instances are concurrent-safe (possibly raising a warning or prompt to continue).

The downside to leaving this to a HOT is that any tool that is passing a _configuration document_ to DSC, rather than using the DSC CLI or API to invoke resources, will struggle to provide this functionality.

This was much more neatly addressable with the design space of group resources, but given the shift away from nested instances due to the difficulties for parsing expressions, that's not as much of an option.

If we plan to continue to support group resources, I would much prefer to see this functionality implemented that way, with a resource that is either a recursive DSC invocation or built on top of `dsc-lib`. That would enable very explicit contracts and clarity that maps what DSC will do when processing the configuration to the text in the configuration document in a way that hoisting concurrent-safe resources to a parallel process doesn't.

**Comment by theJasonHelmick** [comment]
Thanks @SteveL-MSFT for raising this — we discussed parallelism in the DSC working group.

**Current direction:**

- **DSC will not implement parallel execution in the engine.**  
  Determining whether resources can safely run in parallel is not something the engine can reliably validate. Introducing concurrency at this layer would risk non-deterministic behavior and make configurations harder to reason about and troubleshoot.

- **Parallelism is expected to be handled by higher-order tools.**  
  Tools orchestrating DSC (for example, deployment systems or configuration generators) are in a better position to control concurrency, sequencing, and dependency management.

- **We do see value in resource-level metadata for concurrency.**  
  We plan to introduce a field in the resource manifest where authors can declare that multiple instances of the same resource are safe to run in parallel. This is an advisory signal:
  - It only applies to the resource itself  
  - It does not account for interactions with other resources or the broader system  
  - The caller (tooling or user) is responsible for using it correctly  

In short, DSC will remain deterministic and sequential at the engine level, while enabling higher-level tooling to take advantage of parallelism where appropriate.

We’ll track the resource manifest metadata as a separate follow-up issue here: https://github.com/PowerShell/DSC/issues/1459
